### DISCUSSION TITLE
can i jump between master and develop branch?

### DISCUSSION THREAD

**Discussion by molleweide** [NTBBloodbath/doom-nvim#85] [Q&A] [answered]
Can I switch between master branch and develop branch without any problems.
Is there anything that i need to be cautious about?? I want to follow the development branch so
that i am more up to date.

**Comment by NTBBloodbath** [comment, marked as answer]
Hi, yes, you can simply track the remote develop branch and then switch it locally and there shouldn't be any issue _except for the packer stuff_.

If you get some errors with plugins stuff just remove the `plugin/packer_compiled.lua` and doom should regenerate it in the next startup

**Comment by molleweide** [comment]
aaaah thanks for the rm packer comp tip!!!

**Comment by molleweide** [comment]
also i saw this one. it looks realy cool. what do you think?

https://github.com/windwp/windline.nvim

**Comment by NTBBloodbath** [comment]
I'll need to see how extensible is it compared with galaxyline. However if it doesn't takes my attention after reading the source then you can disable doom's default statusline and setup windline by yourself :thinking:

**Comment by molleweide** [comment]
nice. i just wanted to show you. I have to get down with the dev branch now so that I can try out more stuff and make PRs and add value to the project. 

the reason why i though windline was cool is because i often cannot see which window split i am working in and so his color thingys seem to make that quite clear. i haven't tried it myself yet tho. :D

**Comment by molleweide** [comment]
ey are the config and doom rc supposed to be empty in develop?

**Comment by NTBBloodbath** [comment]
> ey are the config and doom rc supposed to be empty in develop?

Nope, they shouldn't. How did you jumped into the develop branch? :thinking:

**Comment by molleweide** [comment]
i did `git fetch upstream` and then `checkout upstream develop` or `upstream/develop`

**Comment by molleweide** [comment]
also, if i remove packer compiled file it does not autoredo everything for me. i run packer install manually.

**Comment by NTBBloodbath** [comment]
> i did `git fetch upstream` and then `checkout upstream develop` or `upstream/develop`

did you touched something else? This is a really weird behavior for me because I often switch between both branches and these files shouldn't be empty, git did some black magic :confused:

**Comment by molleweide** [comment]
however, i am happy now because now i successfully managed to switch between my own personal branch and debvelop
and after running some packer commands both seem to work in the end.

**Comment by molleweide** [comment]
no 

> did you touched something else? This is a really weird behavior for me because these files shouldn't be empty, git did some black magic 😕

no not that i know of. but i will try to recall everything i did now and post!!

**Comment by molleweide** [comment]
we'll probably figure it out. i always do stupid shit.

**Comment by NTBBloodbath** [comment]
> also, if i remove packer compiled file it does not autoredo everything for me. i run packer install manually.

Hmm maybe it was because the configuration files were cleaned and then doom entered in panic :thinking:

**Comment by molleweide** [comment]
haha

**Comment by molleweide** [comment]
i have to take a break now but i post back later

**Comment by NTBBloodbath** [comment]
> no
> 
> > did you touched something else? This is a really weird behavior for me because these files shouldn't be empty, git did some black magic confused
> 
> no not that i know of. but i will try to recall everything i did now and post!!

sure thing, let me know if you find something :slightly_smiling_face: 

in the worst case the only thing left will be having to configure again :thinking:

**Comment by molleweide** [comment]
i installed to custom dir.....
that's why it complains. everything probably works. i just have to go in and change the doom root dir. 
where do i do that in the develop branch.

i installed my nvim configs to ~/.config/nvim.config/doom and then I use cheovim in ~/.config/nvim for selecting which config I want from nvim.config.

**Comment by molleweide** [comment]
I see that have rearranged stuff quit a lot. I am looking into the utils.lua file now but I am going to have to sit down and do a more thorough look later because there more to consider than i thought. i mean, before, i just replaced the doom root variable but now there's more stuff going on. but this is a minor problem which is good.

**Comment by molleweide** [comment]
dude now it works. it is amazing!!!!

**Comment by molleweide** [comment]
i get this packer err on vim load

```
Error running config for nvim-mapper
Error in packer_compiled: ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:52: [string "..."]:0: loop or previous error loading module 'doo
m.utils'
Please check your config for correctness
Press ENTER or type command to continue
```

**Comment by molleweide** [comment]
also do you want me to make a new issue for each specific "minor" error that i stumble upon. because I'd love to do that unless you hate me for spamming issues. ofc i'm half joking but you know what i mean :P

**Comment by molleweide** [comment]
also last question for now i believe. 

How do I as a newbie  put together a pull request of all my changes if i have also very personal keymappings intertwined amongst my commits. I would love to know what is the most pro way of doing this so that I can create solid pr's in the future. would you mind outlining a short list of the commands that I should use and I will read up on all of the on my own and then create a commit of my jump commands that i have created and a couple of other ones that might be useful.

**Comment by molleweide** [comment]
i found this one!!

https://stackoverflow.com/questions/12660839/pull-request-for-only-certain-files-commits

i will split commits to new branch

**Comment by molleweide** [comment]
Another thing that is confusing me is why this doesn't work.

```
utils.map( 'n', '<leader>kjy', '<cmd>e ~/.dotfiles/yabairc<CR>', opts )
utils.map( 'n', '<leader>kjza', '<cmd>e ~/.dotfiles/zsh/custom/aliases.zsh<CR>', opts )
utils.map( 'n', '<leader>kjze', '<cmd>e ~/.dotfiles/zsh/custom/exports.zsh<CR>', opts )
utils.map( 'n', '<leader>kjzz', '<cmd>e ~/.dotfiles/zshrc<CR>', opts)
```
all of these worked in the main branch but not under dev.

**Comment by molleweide** [comment]
also when i said that i am a newbie with git I mean PRs explicitly. rebase and doing voodoo with commits and stuff. I don't want you to think that i am more stupid than i am lol

**Comment by NTBBloodbath** [comment]
> I see that have rearranged stuff quit a lot. I am looking into the utils.lua file now but I am going to have to sit down and do a more thorough look later because there more to consider than i thought. i mean, before, i just replaced the doom root variable but now there's more stuff going on. but this is a minor problem which is good.

Yup but you only need to care about the `doom_root` and `doom_configs_root` variables because these are the variables that handles how everything works internally :)

Also sorry for the very late reply, I was sleeping haha

**Comment by NTBBloodbath** [comment]
> also do you want me to make a new issue for each specific "minor" error that i stumble upon. because I'd love to do that unless you hate me for spamming issues. ofc i'm half joking but you know what i mean :P

I prefer a new issue for each specific error, in that way we'll have a better organization :)

**Comment by NTBBloodbath** [comment]
> also last question for now i believe.
> 
> How do I as a newbie put together a pull request of all my changes if i have also very personal keymappings intertwined amongst my commits. I would love to know what is the most pro way of doing this so that I can create solid pr's in the future. would you mind outlining a short list of the commands that I should use and I will read up on all of the on my own and then create a commit of my jump commands that i have created and a couple of other ones that might be useful.

There's a pretty good guide for PRs mentioned in our [contributing](https://github.com/NTBBloodbath/doom-nvim/blob/main/docs/contributing.md#contributing) docs section, hope it helps!

**Comment by NTBBloodbath** [comment]
> Another thing that is confusing me is why this doesn't work.
> 
> ```
> utils.map( 'n', '<leader>kjy', '<cmd>e ~/.dotfiles/yabairc<CR>', opts )
> utils.map( 'n', '<leader>kjza', '<cmd>e ~/.dotfiles/zsh/custom/aliases.zsh<CR>', opts )
> utils.map( 'n', '<leader>kjze', '<cmd>e ~/.dotfiles/zsh/custom/exports.zsh<CR>', opts )
> utils.map( 'n', '<leader>kjzz', '<cmd>e ~/.dotfiles/zshrc<CR>', opts)
> ```
> 
> all of these worked in the main branch but not under dev.

Ahh yes, that's expected if you are using the `nvim-mapper` plugin because it defines some extra fields for our `utils.map()` function. You should use `vim.api.nvim_set_keymap()` function instead if you don't want to add your keybindings to `nvim-mapper`.

Otherwise if you want to add them, the extra fields that you're missing are a `category`, `unique_id` and `description`, they should be declared at the tail of the function arguments

**Comment by NTBBloodbath** [comment]
> i get this packer err on vim load
> 
> ```
> Error running config for nvim-mapper
> Error in packer_compiled: ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:52: [string "..."]:0: loop or previous error loading module 'doom.utils'
> Please check your config for correctness
> Press ENTER or type command to continue
> ```

Ah, I introduced this one yesterday. I'll fix it asap :)

**Comment by molleweide** [comment]
yo i merge upstream dev again and it crashes with this error

```
Error detected while processing /Users/hjalmarjakobsson/.dotfiles/nvim/init.lua:
E5113: Error while calling lua chunk: .../site/pack/cheovim/start/cheovim/lua/doom/utils/init.lua:13: attempt to index global 'system' (a nil value)
Press ENTER or type command to continue
```

**Comment by NTBBloodbath** [comment]
Hey, you should pull the changes again, `utils/init.lua` doesn't have any reference to `system` module right now :thinking:

**Comment by NTBBloodbath** [comment]
Also I moved the `doom_root` and `doom_configs_root` variables to [lua/doom/core/systen/init.lua](https://github.com/NTBBloodbath/doom-nvim/blob/develop/lua/doom/core/system/init.lua) btw

**Comment by molleweide** [comment]
is utils used at all?? i mean it seems to be duplicate stuff?

**Comment by NTBBloodbath** [comment]
the utils module doesn't manage the paths stuff anymore 🤔

**Comment by molleweide** [comment]
great because i had already done a lot of stupid shit based on utils lol

**Comment by molleweide** [comment]
fml there are too many errors now haha and when i switch back to previous working state i get this below.
i apologize for bothering so much but i really want to get the dev branch to work so that i can contribute to it :P

```
Error running config for telescope.nvim
E5108: Error executing lua ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:115: Vim(echomsg):E114: Missing quote: "Error in packer_compile
d: ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:52: ...er/opt/telescope.nvim/lua/telescope/_extensions/init.lua:11: This extension does
n't exist or is not installed: mapper
Press ENTER or type command to continue
```

is there an easy fix for this. i have tried packercompile and paker clean/install

**Comment by NTBBloodbath** [comment]
> great because i had already done a lot of stupid shit based on utils lol

Ye I've been playing around with some things and refactoring code today but the most important changes are in [this](https://github.com/NTBBloodbath/doom-nvim/commit/e976bcae2cb931bb4f410b4236016318ade7a8b2#diff-319c75de4d287d92da3916501025d7276f16acc8277664d733e5aa9f585a2d5c) commit

**Comment by NTBBloodbath** [comment]
> fml there are too many errors now haha and when i switch back to previous working state i get this
> 
> ```
> Error running config for telescope.nvim
> E5108: Error executing lua ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:115: Vim(echomsg):E114: Missing quote: "Error in packer_compile
> d: ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:52: ...er/opt/telescope.nvim/lua/telescope/_extensions/init.lua:11: This extension does
> n't exist or is not installed: mapper
> Press ENTER or type command to continue
> ```

Try removing the `plugin/packer_compiled.lua` file, this one can be a headache some times lol

**Comment by molleweide** [comment]
> > great because i had already done a lot of stupid shit based on utils lol
> 
> Ye I've been playing around with some things and refactoring code today but the most important changes are in [this](https://github.com/NTBBloodbath/doom-nvim/commit/e976bcae2cb931bb4f410b4236016318ade7a8b2#diff-319c75de4d287d92da3916501025d7276f16acc8277664d733e5aa9f585a2d5c) commit

ooh i will read through this!!!

also maybe it would make sense to have a variable called something like `doom_custom_path` that you append at the end of the doom_root variables so that all custom paths within ~/.config will work so that this is the only on one needs to modify if one uses a custom dir. but I should read through the commits you provided first before i recommend too much..

or no actually i think my above reasoning makes sense so that you can always only modify one place if you have some custom path. I chose to use ~/.config/nvim.config/doom because I have never worked with multiple configs and so maybe there is a much bette way of doing this but it felt wierd for me to put doom-nvim in config root since it is an nvim config that is why i wanted to keep all my nvim configs in one dir. is this bad thinking? won't the config dir be clogged with too many configs if you put all of them in XDG config root?

**Comment by molleweide** [comment]
> > fml there are too many errors now haha and when i switch back to previous working state i get this
> > ```
> > Error running config for telescope.nvim
> > E5108: Error executing lua ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:115: Vim(echomsg):E114: Missing quote: "Error in packer_compile
> > d: ...vim/site/pack/packer/opt/packer.nvim/lua/packer/load.lua:52: ...er/opt/telescope.nvim/lua/telescope/_extensions/init.lua:11: This extension does
> > n't exist or is not installed: mapper
> > Press ENTER or type command to continue
> > ```
> 
> Try removing the `plugin/packer_compiled.lua` file, this one can be a headache some times lol

yeah luckilly i created a new branch before pulling so i didn't shoot myself completely so i will try to fix all the errors in the pullupstreamdev branch :D

**Comment by molleweide** [comment]
in the end this is very good for my learning but it makes me very pissed in the moment

**Comment by molleweide** [comment]
i believe XDG config is hardcoded in init.lua still

**Comment by NTBBloodbath** [comment]
> > > great because i had already done a lot of stupid shit based on utils lol
> > 
> > 
> > Ye I've been playing around with some things and refactoring code today but the most important changes are in [this](https://github.com/NTBBloodbath/doom-nvim/commit/e976bcae2cb931bb4f410b4236016318ade7a8b2#diff-319c75de4d287d92da3916501025d7276f16acc8277664d733e5aa9f585a2d5c) commit
> 
> ooh i will read through this!!!
> 
> also maybe it would make sense to have a variable called something like `doom_custom_path` that you append at the end of the doom_root variables so that all custom paths within ~/.config will work so that this is the only on one needs to modify if one uses a custom dir. but I should read through the commits you provided first before i recommend too much..

This is not needed IMO, this is the function of our `doom_root` variable.

Basically this is how our path works:

1. Everything goes to where `doom_root` variable points to, it's like the reference for the core path of doom.
2. The `doom_configs_root` was created very recently with the purpose of allowing to version control your Doom Nvim configurations without having to create a submodule in your dotfiles, you can see more about how it works in #79
3. Because of the point 2, we can have some conflicts if we have our configurations in other path and doom can't find them where `doom_root` points to, this is why we also set a fallback path for the configurations location in the configuration modules, based on `doom_root` and `doom_configs_root`.

Hope this was understandable (I feel like I'm somewhat bad explaining lmao) and useful :)

**Comment by molleweide** [comment]
is this done correctly then? this worked previously for me but I want to make sure this will work with your philosophy.
```
M.doom_configs_root = string.format("%s%snvim.config/doom", M.config_dir, M.sep)
```

**Comment by NTBBloodbath** [comment]
> i believe XDG config is hardcoded in init.lua still

Yeah, this is because the `XDG_CONFIG_HOME` is an environment variable and its value can be changed by the end user in his bashrc, zshrc, etc

**Comment by NTBBloodbath** [comment]
> okay so is this how i should modify system for myself:
> 
> ```
> M.doom_configs_root = string.format("%s%snvim.config/doom", M.config_dir, M.sep)
> ```

Yup, that's how it should look like :)

**Comment by molleweide** [comment]
> > i believe XDG config is hardcoded in init.lua still
> 
> Yeah, this is because the `XDG_CONFIG_HOME` is an environment variable and its value can be changed by the end user in his bashrc, zshrc, etc

but this doesn't work unless you have doom in config root righ so i have to modify this?

**Comment by molleweide** [comment]
woaw now it all works all of a sudden lol... i think!!! 🤔 😆

**Comment by NTBBloodbath** [comment]
Yes, if you have doom inside other path e.g. `~/Development/Nvim` you will need to adapt `doom_root` to point to this path

**Comment by NTBBloodbath** [comment]
Btw maybe we should convert this into a discussion 🤔

**Comment by molleweide** [comment]
> Btw maybe we should convert this into a discussion 🤔

yes indeed i have no idead how to do that tho

**Comment by NTBBloodbath** [comment]
Lemme do it, hold on

**Comment by NTBBloodbath** [comment]
Done :)

**Comment by molleweide** [comment]
yo!
have you made any changes to keymap/init or which key config since last develop push. if not i will checkout these files from my previous develop branch.

**Comment by NTBBloodbath** [comment]
I did some additions for the DAP plugins 3 days ago in the whichkey setup and the keybindings/init had some small modifications yesterday too

**Comment by molleweide** [comment]
cool then it shouldn't be a problem because i workd on this late in the evning yesterday

**Comment by NTBBloodbath** [comment]
Let's hope it :)

**Comment by molleweide** [comment]
allright so i have to now read up on how to make a good pr and then push. the jump commands became really good

**Comment by NTBBloodbath** [comment]
I'm pretty sure that they'll be really useful 🙂 

Btw you'll surely want to read [this](https://github.com/NTBBloodbath/doom-nvim/blob/51315be63bf0f0a2aa3ebe635b9c75d4830e0ca2/docs/contributing.md#contributing-code), it's basically the doom's code conventions and where should be placed some stuff for a consistent code :)

**Reply by molleweide** [reply]
Nice i really have to become more comfy w rebasing etc

**Comment by molleweide** [comment]
hey man are you using something like this or how do you accesss manpages in vim

https://github.com/jez/vim-superman

**Reply by NTBBloodbath** [reply]
not really lol, never used these things tbh however it would be pretty cool to have something for it :)

**Comment by molleweide** [comment]
i also decided to put doom-nvim back under XDG_HOME so that it becomes less of a hassle to test before pushing the pr.

**Comment by molleweide** [comment]
hey have you tried NvChad also lol??

I'm gonna see if i can get it to work with cheovim and take inspiration from it later.

**Reply by NTBBloodbath** [reply]
Yeah, but haven't used it in a while (I'm too lazy to update it lmao). NvChad has good stuff too imo

**Comment by molleweide** [comment]
I am also doing a deep walkthrough of the doom_config file now and I see that I wasn't aware of some nice track etc. in it so my apologize for some of the stupid question I have been posting earlier. Especially for not having set the log levels.

**Reply by NTBBloodbath** [reply]
No worries man, always happy to help!
